In [1]:
import pandas as pd
df=pd.read_csv('C:/Users/DELL/OneDrive/Documents/data science asignments/EDA2/EDA2/adult_with_headers.csv')

In [2]:
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

In [3]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
income            object
dtype: object

In [5]:
df.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [8]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [9]:
import pandas as pd

# Load your dataset
file_path = 'C:/Users/DELL/OneDrive/Documents/data science asignments/EDA2/EDA2/adult_with_headers.csv'
df = pd.read_csv(file_path)

# Remove extra spaces and replace "?" with NaN across the entire DataFrame
df = df.applymap(lambda x: pd.NA if str(x).strip() == "?" else x)

# Now, check for missing values
missing_values = df.isnull().sum()
print("Missing values per column:\n", missing_values)


Missing values per column:
 age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     583
income               0
dtype: int64


C:\Users\DELL\AppData\Local\Temp\ipykernel_17708\1987929927.py:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: pd.NA if str(x).strip() == "?" else x)


In [10]:
df.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     583
income               0
dtype: int64

In [13]:
import pandas as pd

# Load the dataset and check if it's loaded correctly
file_path ='C:/Users/DELL/OneDrive/Documents/data science asignments/EDA2/EDA2/adult_with_headers.csv' 
df = pd.read_csv(file_path)

# Replace '?' with NaN
df = df.replace('?', pd.NA)

# Calculate the mode and fill missing values directly
workclass_mode = df['workclass'].mode()[0]
df['workclass'] = df['workclass'].fillna(workclass_mode)

occupation_mode = df['occupation'].mode()[0]
df['occupation'] = df['occupation'].fillna(occupation_mode)

native_country_mode = df['native_country'].mode()[0]
df['native_country'] = df['native_country'].fillna(native_country_mode)

# Confirm no missing values remain in these columns
print(df[['workclass', 'occupation', 'native_country']].isnull().sum())


workclass         0
occupation        0
native_country    0
dtype: int64


In [14]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [17]:
processed_file_path = 'C:/Users/DELL/OneDrive/Documents/data science asignments/EDA2/EDA2/adultt_processed.csv'
df.to_csv(processed_file_path, index=False)

print(f"Processed dataset saved to: {processed_file_path}")


Processed dataset saved to: C:/Users/DELL/OneDrive/Documents/data science asignments/EDA2/EDA2/adultt_processed.csv


Standard Scaling
Standard scaling (Z-score normalization) transforms the data to have a mean of 0 and a standard deviation of 1.

In [18]:

from sklearn.preprocessing import StandardScaler

# Select numerical features
numerical_features = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

# Initialize the scaler
scaler = StandardScaler()

# Apply standard scaling
df[numerical_features] = scaler.fit_transform(df[numerical_features])

print("Standard Scaled Data:")
print(df[numerical_features].head())


Standard Scaled Data:
        age    fnlwgt  education_num  capital_gain  capital_loss  \
0  0.030671 -1.063611       1.134739      0.148453      -0.21666   
1  0.837109 -1.008707       1.134739     -0.145920      -0.21666   
2 -0.042642  0.245079      -0.420060     -0.145920      -0.21666   
3  1.057047  0.425801      -1.197459     -0.145920      -0.21666   
4 -0.775768  1.408176       1.134739     -0.145920      -0.21666   

   hours_per_week  
0       -0.035429  
1       -2.222153  
2       -0.035429  
3       -0.035429  
4       -0.035429  


In [ ]:
Min-Max Scaling
Min-Max scaling transforms the data to a fixed range, usually [0, 1].

In [19]:
from sklearn.preprocessing import MinMaxScaler

# Initialize the scaler
min_max_scaler = MinMaxScaler()

# Apply min-max scaling
df[numerical_features] = min_max_scaler.fit_transform(df[numerical_features])

print("Min-Max Scaled Data:")
print(df[numerical_features].head())

Min-Max Scaled Data:
        age    fnlwgt  education_num  capital_gain  capital_loss  \
0  0.301370  0.044302       0.800000       0.02174           0.0   
1  0.452055  0.048238       0.800000       0.00000           0.0   
2  0.287671  0.138113       0.533333       0.00000           0.0   
3  0.493151  0.151068       0.400000       0.00000           0.0   
4  0.150685  0.221488       0.800000       0.00000           0.0   

   hours_per_week  
0        0.397959  
1        0.122449  
2        0.397959  
3        0.397959  
4        0.397959  


Analyzing the outputs of both Standard Scaling and Min-Max Scaling can provide insights into how these techniques transform the data and their implications for subsequent analysis or modeling. Here’s a detailed analysis based on the provided scaled data:
Overview of Scaled Data
Standard Scaled Data:
The values are centered around 0 with a standard deviation of 1.
This scaling method is particularly useful when the features are normally distributed or when using algorithms sensitive to the scale of input data.
Min-Max Scaled Data:
The values are transformed to a range between 0 and 1.
This approach is beneficial when you want to ensure that all features contribute equally to distance calculations, particularly in algorithms like K-nearest neighbors (KNN) or neural networks.
Analysis of Both Scaling Techniques
1. Standard Scaling Analysis
Data Distribution:
The output shows that after standard scaling, values can be both positive and negative. For instance, an age value of 0.030671 indicates it is slightly above the mean, while -0.775768 suggests it is below average.
Implications:
Models that assume normally distributed data (like linear regression or logistic regression) can benefit from this scaling because it maintains the relationships between features while normalizing their distributions.
Outlier Influence:
While standard scaling does not eliminate outliers, it reduces their influence compared to Min-Max scaling, which can compress the range of non-outlier values.
2. Min-Max Scaling Analysis
Range Transformation:
All values are confined within the range of [0, 1]. For example, an age value of 0.301370 indicates that this individual falls within a certain percentile of the dataset.
Implications:
This scaling technique is particularly useful for algorithms that require bounded input, such as neural networks with activation functions like sigmoid or tanh.
Sensitivity to Outliers:
Min-Max scaling is sensitive to outliers because it uses the minimum and maximum values for scaling. If there are extreme values in the dataset, they can significantly distort the scaled values of other observations.
Comparison of Outputs
Feature	Standard Scaled Value	Min-Max Scaled Value
Age	0.030671	0.301370
Fnlwgt	-1.063611	0.044302
Education_num	1.134739	0.800000
Capital_gain	0.148453	0.021740
Capital_loss	-0.216660	0.000000
Hours_per_week	-0.035429	0.397959
Key Takeaways
Model Selection: Choose Standard Scaling when using models that assume normality in the data distribution or when dealing with algorithms sensitive to feature magnitudes (e.g., SVM, PCA). Opt for Min-Max Scaling when working with algorithms that require bounded inputs or when you need to preserve relationships between features.
Feature Interpretation: After applying standard scaling, you can interpret how far each observation deviates from the mean, which can be useful in understanding feature importance and relationships in models.
Impact on Distance Metrics: In distance-based algorithms (like KNN), Min-Max Scaling ensures that all features contribute equally since they are on the same scale, preventing features with larger ranges from dominating distance calculations.

One-Hot Encoding and Label Encoding are two common techniques used to convert categorical variables into a numerical format that can be used in machine learning algorithms. Here's a detailed explanation of each method and how they function:
One-Hot Encoding
Definition: One-Hot Encoding transforms categorical variables into a format that can be provided to machine learning algorithms to improve predictions. It creates binary (0 or 1) columns for each category in the original variable.
How It Works:
For each unique category in the categorical variable, a new binary column is created.
Each observation is marked with a 1 in the column corresponding to its category and 0 in all other new columns.
Example:
Suppose you have a categorical variable Color with three categories: Red, Green, and Blue.
Color
Red
Green
Blue
Green
Red
Using One-Hot Encoding, this would be transformed into:
Color_Red	Color_Green	Color_Blue
1	0	0
0	1	0
0	0	1
0	1	0
1	0	0
Label Encoding
Definition: Label Encoding converts categorical variables into numerical values by assigning each unique category an integer value. This method is straightforward but can introduce unintended ordinal relationships between categories.
How It Works:
Each unique category is assigned a unique integer starting from 0.
The original categorical variable is replaced with these integer values.
Example:
Using the same Color variable:
Color
Red
Green
Blue
Green
Red
Using Label Encoding, this would be transformed into:
Color (Encoded)
2
1
0
1
2
Key Differences
Output Format:
One-Hot Encoding results in multiple binary columns for each category.
Label Encoding results in a single column with integer values.
Interpretation of Values:
In One-Hot Encoding, there is no ordinal relationship implied; each category is independent.
In Label Encoding, the integer values imply an order, which may not exist between categories.
When to Use Each Method
One-Hot Encoding is preferred when:
The categorical variable does not have an ordinal relationship.
The number of unique categories is relatively small (to avoid high dimensionality).
Label Encoding is preferred when:
The categorical variable has an ordinal relationship (e.g., ratings like "low," "medium," "high").
The number of unique categories is large, and creating many binary columns would lead to excessive dimensionality.
Conclusion
Both One-Hot Encoding and Label Encoding are essential techniques for preparing categorical data for machine learning. The choice between them depends on the nature of the data and the specific requirements of the algorithms being used. Understanding how they work helps ensure that your models are trained effectively without introducing bias or misinterpretation of the data.


In [24]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

# Load the dataset


# Replace '?' with NaN for proper handling of missing values
df.replace('?', pd.NA, inplace=True)

# Drop rows with missing values (or handle them as needed)
df.dropna(inplace=True)

# Initialize encoders
one_hot_encoder = OneHotEncoder(sparse_output=False)  # Updated parameter name
label_encoder = LabelEncoder()

# Apply One-Hot Encoding to categorical variables with < 5 categories
one_hot_columns = ['sex', 'race', 'relationship']
one_hot_encoded = one_hot_encoder.fit_transform(df[one_hot_columns])
one_hot_df = pd.DataFrame(one_hot_encoded, columns=one_hot_encoder.get_feature_names_out(one_hot_columns))

# Apply Label Encoding to categorical variables with > 5 categories
label_columns = ['workclass', 'education', 'marital_status', 'occupation', 'native_country']
for col in label_columns:
    df[col] = label_encoder.fit_transform(df[col])

# Combine the original DataFrame with the one-hot encoded DataFrame
df_encoded = pd.concat([df.drop(one_hot_columns, axis=1), one_hot_df], axis=1)

print(df_encoded.head())

        age  workclass    fnlwgt  education  education_num  marital_status  \
0  0.301370        5.0  0.044302        9.0       0.800000             4.0   
1  0.452055        4.0  0.048238        9.0       0.800000             2.0   
2  0.287671        2.0  0.138113       11.0       0.533333             0.0   
3  0.493151        2.0  0.151068        1.0       0.400000             2.0   
4  0.150685        2.0  0.221488        9.0       0.800000             2.0   

   occupation  capital_gain  capital_loss  hours_per_week  ...  \
0         0.0       0.02174           0.0        0.397959  ...   
1         3.0       0.00000           0.0        0.122449  ...   
2         5.0       0.00000           0.0        0.397959  ...   
3         5.0       0.00000           0.0        0.397959  ...   
4         9.0       0.00000           0.0        0.397959  ...   

   race_Asian-Pac-Islander race_Black  race_Other  race_White  \
0                      0.0        0.0         0.0         1.0   
1   

. Feature Engineering:
•	Create at least 2 new features that could be beneficial for the model. Explain the rationale behind your choices.
•	Apply a transformation (e.g., log transformation) to at least one skewed numerical feature and justify your choice.
To enhance the dataset through feature engineering, we will create two new features and apply a transformation to one skewed numerical feature. Below are the details of the new features and the rationale behind their creation, as well as the transformation applied to a numerical feature.
New Features Creation
1. Income Binary Feature
Feature Name: income_binary
Description: This feature will convert the income categorical variable into a binary format (0 for <=50K and 1 for >50K).
Rationale: Many machine learning algorithms perform better with numerical inputs. By converting income into a binary format, we can simplify the model's task of predicting income levels based on other features.
2. Experience Level
Feature Name: experience_level
Description: This feature will be derived from the education_num and age columns to estimate an individual’s experience level. It can be calculated as:
experience level
=
age
−
education num
−
6
experience level=age−education num−6
Here, we subtract 6 years, assuming that formal education typically starts at age 6.
Rationale: This feature could provide insights into how age and education interact to influence income potential. A higher experience level may correlate with higher income.
Applying Transformations
Log Transformation on fnlwgt
Feature Name: fnlwgt
Transformation: Apply a log transformation.
Rationale: The fnlwgt (final weight) feature is often skewed, meaning it has a long tail of high values. Log transformation helps in normalizing this skewness, making it more suitable for linear models that assume normally distributed data. It can also stabilize variance and make the data more interpretable.


In [26]:
import pandas as pd
import numpy as np

# Load the dataset


# Replace '?' with NaN for proper handling of missing values
df.replace('?', pd.NA, inplace=True)

# Drop rows with missing values (or handle them as needed)
df.dropna(inplace=True)

# Create binary income feature
df['income_binary'] = df['income'].apply(lambda x: 1 if x == '>50K' else 0)

# Create experience level feature
df['experience_level'] = df['age'] - df['education_num'] - 6

# Apply log transformation to fnlwgt
df['fnlwgt_log'] = np.log(df['fnlwgt'])

# Display the updated DataFrame with new features
print(df[['income_binary', 'experience_level', 'fnlwgt_log']].head())

   income_binary  experience_level  fnlwgt_log
0              0         -6.498630   -3.116728
1              0         -6.347945   -3.031617
2              0         -6.245662   -1.979680
3              0         -5.906849   -1.890023
4              0         -6.649315   -1.507385


By creating the income_binary and experience_level features, we enhance the dataset's potential for predictive modeling. The log transformation applied to fnlwgt addresses skewness, making it more suitable for analysis. These modifications aim to improve model performance and interpretability by providing additional insights into relationships within the data.


Feature Selection: Outlier Detection and Predictive Power Score
In this section, we will use the Isolation Forest algorithm to identify and remove outliers from the dataset, followed by applying the Predictive Power Score (PPS) to analyze relationships between features. We will also compare the findings from PPS with those from a correlation matrix.
1. Outlier Detection Using Isolation Forest
Isolation Forest is an effective algorithm for detecting outliers in high-dimensional datasets. It works by isolating observations through random partitioning, where outliers are easier to isolate compared to normal observations.
Implementation of Isolation Forest

In [27]:
import pandas as pd
from sklearn.ensemble import IsolationForest

# Load the dataset


# Replace '?' with NaN for proper handling of missing values
df.replace('?', pd.NA, inplace=True)

# Drop rows with missing values (or handle them as needed)
df.dropna(inplace=True)

# Select numerical features for outlier detection
numerical_features = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

# Initialize Isolation Forest
iso_forest = IsolationForest(contamination=0.05)  # Assuming 5% of data may be outliers

# Fit the model and predict outliers
outlier_predictions = iso_forest.fit_predict(df[numerical_features])

# Add predictions to DataFrame: -1 for outliers, 1 for inliers
df['outlier'] = outlier_predictions

# Remove outliers from the dataset
df_cleaned = df[df['outlier'] != -1]

print(f"Original dataset size: {df.shape[0]}")
print(f"Cleaned dataset size: {df_cleaned.shape[0]}")

Original dataset size: 30162
Cleaned dataset size: 28653


Discussion: How Outliers Affect Model Performance
Bias in Predictions: Outliers can skew the results of statistical models, leading to biased estimates and predictions. For example, linear regression can be heavily influenced by extreme values, resulting in a model that does not generalize well to new data.
Increased Variance: Models may become overly complex if they try to fit to outliers, leading to overfitting. This reduces the model's ability to perform well on unseen data.
Impact on Metrics: Outliers can distort performance metrics such as mean squared error (MSE), making it difficult to assess model performance accurately.

. Predictive Power Score (PPS)
The Predictive Power Score (PPS) quantifies the predictive power of one feature over another, providing insights into their relationships. It ranges from 0 (no predictive power) to 1 (perfect predictive power).
Implementation of PPS

In [33]:
import ppscore as pps

# Calculate PPS for all feature pairs
pps_matrix = pps.matrix(df_cleaned)

# Display the PPS matrix
print(pps_matrix)

ModuleNotFoundError: No module named 'ppscore'

In [32]:
correlation_matrix = df_cleaned[numerical_features].corr()
print(correlation_matrix)

                     age    fnlwgt  education_num  capital_gain  capital_loss  \
age             1.000000 -0.082777       0.044908      0.088587      0.034414   
fnlwgt         -0.082777  1.000000      -0.041352     -0.016057     -0.016374   
education_num   0.044908 -0.041352       1.000000      0.138579      0.029878   
capital_gain    0.088587 -0.016057       0.138579      1.000000     -0.031959   
capital_loss    0.034414 -0.016374       0.029878     -0.031959      1.000000   
hours_per_week  0.119033 -0.020885       0.136192      0.073250      0.022205   

                hours_per_week  
age                   0.119033  
fnlwgt               -0.020885  
education_num         0.136192  
capital_gain          0.073250  
capital_loss          0.022205  
hours_per_week        1.000000  


Discussion: Comparing PPS and Correlation Matrix Findings
Nature of Relationships:
Correlation Matrix: Only captures linear relationships. For example, a high positive correlation between age and education_num indicates that older individuals tend to have higher education levels.
PPS: Captures both linear and non-linear relationships. It can reveal complex interactions that might not be evident in the correlation matrix.
Interpretation:
The correlation matrix might show a strong relationship between two features, but PPS could indicate that one feature has little predictive power over another when considering other variables.
Outlier Influence:
The correlation matrix can be significantly influenced by outliers, while PPS provides a more robust measure of predictive power that is less affected by extreme values.